File: Algorithm/Leetcode322.cpp
/* Leetcode 322 Coin Change */
//给定不同面额的硬币 coins 和一个总金额 amount。编写一个函数来计算可以凑成总金额所需的最少的硬币个数。
//如果没有任何一种硬币组合能组成总金额，返回 -1。

#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
	int coinChange(vector<int>& coins, int amount) {
		const int MAX = 0x3f3f3f3f;
		vector<int> dp(amount + 1, MAX);
		int n = coins.size();
		dp[0] = 0;
		for (int i = 1; i <= amount; i++) {
			for (int j = 0; j < n; j++) {
				if (i >= coins[j]) {
					dp[i] = min(dp[i], dp[i - coins[j]] + 1);
				}
			}
		}
		if (dp[amount] > amount) {
			return -1;
		}
		return dp[amount];
	}
};

File: Algorithm/DP/LeetCode322.java
import java.util.Arrays;

public class LeetCode322 {
    /* LeetCode 322 零钱兑换 */

    /**
     * 给定不同面额的硬币 coins 和一个总金额 amount。编写一个函数来计算可以凑成总金额所需的最少的硬币个数。
     * 如果没有任何一种硬币组合能组成总金额，返回 -1。
     */
    public static void main(String[] args) {
        Solution solution = new Solution();
    }
}

class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = amount + 1;
        int[] dp = new int[max];
        Arrays.fill(dp, max);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.length; j++) {
                int coin = coins[j];
                if(i >= coin) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount] >= max ? -1 : dp[amount];
    }
}
